### PR TITLE
 EIP-7685 Implementation

### LINKED ISSUE #6387: Prototype & implement EIP-7002: Execution layer triggerable exits
https://eips.ethereum.org/EIPS/eip-7002

### PR DESCRIPTION
**Pull request by rjnrohit**
Fixes Closes Resolves #6387 

## Changes

This PR implements the following EIPs:
- 7002 : https://eips.ethereum.org/EIPS/eip-7002
- 6110 : https://eips.ethereum.org/EIPS/eip-6110
- 7685 : https://eips.ethereum.org/EIPS/eip-7685

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

_Optional. Remove if not applicable._

#### Requires documentation update

- [ ] Yes
- [x] No

_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._

#### Requires explanation in Release Notes

- [x] Yes
- [ ] No

_If yes, fill in the details here. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.Consensus/Processing/BlockExtensions.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ internal static class BlockExtensions
         public static Block CreateCopy(this Block block, BlockHeader header) =>
             block is BlockToProduce blockToProduce
                 ? new BlockToProduce(header, blockToProduce.Transactions, blockToProduce.Uncles, blockToProduce.Withdrawals)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockExtensions.cs:21]: @Demuirgos @rjnrohit check this line carefully:
`                ? new BlockToProduce(header, blockToProduce.Transactions, blockToProduce.Uncles, blockToProduce.Withdrawals)	`

analyze all usages of BlockToProduce constructor

File: src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs [hunk 2]
```diff
@@ -325,6 +328,46 @@ private bool ValidateDeposits(Block block, IReleaseSpec spec, out string? error)
         return true;
     }
 
+    public bool ValidateValidatorExits(Block block, out string error) =>
+        ValidateValidatorExits(block, _specProvider.GetSpec(block.Header), out error);
+
+
+    private bool ValidateValidatorExits(Block block, IReleaseSpec spec, out string error)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs:335]: (Cosmetic) For all errors in this method why not to use BlockErrorMessages class similarly to withdrawals

File: src/Nethermind/Nethermind.Core.Test/Builders/BlockBuilder.cs [hunk 3]
```diff
@@ -309,5 +310,14 @@ public BlockBuilder WithParentBeaconBlockRoot(Hash256? parentBeaconBlockRoot)
             TestObjectInternal.Header.ParentBeaconBlockRoot = parentBeaconBlockRoot;
             return this;
         }
+
+        public BlockBuilder WithValidatorExits(ValidatorExit[]? validatorExits)
+        {
+            TestObjectInternal = TestObjectInternal.WithReplacedBody(
+                TestObjectInternal.Body.WithChangedValidatorExits(validatorExits));
+
+            TestObjectInternal.Header.TxRoot = validatorExits is not null ? ValidatorExitsTrie.CalculateRoot(validatorExits) : null;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Builders/BlockBuilder.cs:319]: Copy-paste error (not TxRoot)

File: src/Nethermind/Nethermind.Core.Test/Builders/TestItem.cs [hunk 4]
```diff
@@ -99,12 +99,12 @@ public static Hash256 KeccakFromNumber(int i)
...
+        public static Deposit DepositA_1Eth = new() { Index = 1, PubKey = PublicKeyA.Bytes, Amount = 1_000_000_000, WithdrawalCredentials = AddressA.Bytes, Signature = KeccakA.Bytes.ToArray() };
+        public static Deposit DepositB_2Eth = new() { Index = 2, PubKey = PublicKeyB.Bytes, Amount = 2_000_000_000, WithdrawalCredentials = AddressB.Bytes, Signature = KeccakB.Bytes.ToArray() };
+        public static Deposit DepositC_3Eth = new() { Index = 3, PubKey = PublicKeyC.Bytes, Amount = 3_000_000_000, WithdrawalCredentials = AddressC.Bytes, Signature = KeccakC.Bytes.ToArray() };
+        public static Deposit DepositD_4Eth = new() { Index = 4, PubKey = PublicKeyD.Bytes, Amount = 4_000_000_000, WithdrawalCredentials = AddressD.Bytes, Signature = KeccakD.Bytes.ToArray() };
+        public static Deposit DepositE_5Eth = new() { Index = 5, PubKey = PublicKeyE.Bytes, Amount = 5_000_000_000, WithdrawalCredentials = AddressE.Bytes, Signature = KeccakE.Bytes.ToArray() };
+        public static Deposit DepositF_6Eth = new() { Index = 6, PubKey = PublicKeyF.Bytes, Amount = 6_000_000_000, WithdrawalCredentials = AddressF.Bytes, Signature = KeccakF.Bytes.ToArray() };
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Builders/TestItem.cs:107]: It will be useful to add TestItems for exits, can be done while writing tests

File: src/Nethermind/Nethermind.Core.Test/Encoding/BlockDecoderTests.cs [hunk 5]
```diff
@@ -86,6 +87,18 @@ public BlockDecoderTests()
...
+                .WithBaseFeePerGas(1)
+                .WithTransactions(transactions)
+                .WithUncles(uncles)
+                .WithWithdrawals(8)
+                .WithBlobGasUsed(ulong.MaxValue)
+                .WithExcessBlobGas(ulong.MaxValue)
+                .WithMixHash(Keccak.EmptyTreeHash)
+                // an empty Deposit array
+                .WithDeposits(0)
+                .WithValidatorExits(new[] { new ValidatorExit(TestItem.AddressA, new byte[48]) })
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Encoding/BlockDecoderTests.cs:101]: Considering copy-paste error in WithValidatorExists such tests shouldn't pass, if they pass it means they don't test what they should

File: src/Nethermind/Nethermind.Core.Test/Encoding/HeaderDecoderTests.cs [hunk 6]
```diff
@@ -156,6 +156,46 @@ public void Can_encode_decode_with_cancun_fields(ulong? blobGasUsed, ulong? exce
...
+    {
+        BlockHeader header = Build.A.BlockHeader
+            .WithTimestamp(ulong.MaxValue)
+            .WithBaseFee(1)
+            .WithWithdrawalsRoot(Keccak.Zero)
+            .WithBlobGasUsed(0)
+            .WithExcessBlobGas(0)
+            .WithParentBeaconBlockRoot(TestItem.KeccakB)
+            .WithDepositsRoot(Keccak.Zero)
+            .WithValidatorExitsRoot(TestItem.KeccakA).TestObject;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Encoding/HeaderDecoderTests.cs:170]: Considering copy-paste error in WithValidatorExists such tests shouldn't pass, if they pass it means they don't test what they should

File: src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs [hunk 7]
```diff
@@ -250,6 +255,9 @@ private void ValidateProcessedBlock(Block suggestedBlock, ProcessingOptions opti
         block.Header.ReceiptsRoot = _receiptsRootCalculator.GetReceiptsRoot(receipts, spec, block.ReceiptsRoot);
         ApplyMinerRewards(block, blockTracer, spec);
         _withdrawalProcessor.ProcessWithdrawals(block, spec);
+
+        ProcessValidatorExits(block, spec);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus/Processing/BlockProcessor.cs:259]: I think it will be better (if not specified by spec) if exists will happen after processing deposits. It shouldn't matter, but better safe than sorry

File: src/Nethermind/Nethermind.Core.Test/Encoding/ValidatorExitDecoderTests.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Nethermind.Core.Test.Encoding;
+
+[TestFixture]
+public class ValidatorExitDecoderTests
+{
+    private static ValidatorExitsDecoder _decoder = new();
+
+    [Test]
+    public void Roundtrip()
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Encoding/ValidatorExitDecoderTests.cs:17]: add roundtrip test with more than one exit (exit array)

File: src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs [hunk 9]
```diff
@@ -27,6 +28,8 @@ public class PayloadAttributes
 
     public Deposit[]? Deposits { get; set; }
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Consensus/Producers/PayloadAttributes.cs:29]: @Demuirgos @rjnrohit are you sure that deposits and exists should be added to payloadAttributes? I don't think so

File: src/Nethermind/Nethermind.Core.Test/Encoding/BlockDecoderTests.cs (hunk 5 above)
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Encoding/BlockDecoderTests.cs:101]: add tests with different edge cases for deposits/exits:
null
0
1
more than 1

File: src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecLoader.cs [hunk 10]
```diff
@@ -414,7 +417,7 @@ private static void LoadGenesis(ChainSpecJson chainSpecJson, ChainSpec chainSpec
         if (withdrawalsEnabled)
             genesisHeader.WithdrawalsRoot = Keccak.EmptyTreeHash;
 
-        if(depositsEnabled)
+        if (depositsEnabled)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecLoader.cs:420]: we have the same check two times, depositsEnabled and isEIP6110Enabled, please cleanup

Cosmetic reaname isEip7002Enabled here to existsEnabled

File: src/Nethermind/Nethermind.State/Proofs/ValidatorExitsTrie.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.State/Proofs/ValidatorExitsTrie.cs:4]: For this class for withdrawals we have matching test class:
WithdrawalsTrie ->WithdrawalsTrieTests

Please apply the same for deposits and exits

**Review by MarekM25 — CHANGES_REQUESTED** [review]
Please cleanup, details in comments

**Comment by MarekM25** [comment]
@rjnrohit please fill PR descriptions
image

File: src/Nethermind/Nethermind.Core.Test/Builders/BlockBuilder.cs (hunk 3 above)
Comment (rjnrohit) [reply src/Nethermind/Nethermind.Core.Test/Builders/BlockBuilder.cs:319]: changed! thanks

**Comment by rjnrohit** [comment]
> @rjnrohit please fill PR descriptions image

done, thanks!

File: src/Nethermind/Nethermind.Serialization.Rlp/ConsensusRequestDecoder.cs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            rlpStream.ReadByte();
+            return null;
+        }
+
+        rlpStream.ReadSequenceLength();
+
+        ConsensusRequestsType consensusRequestsType = (ConsensusRequestsType)rlpStream.ReadByte();
+
+        Console.WriteLine("ConsensusRequestsType: " + consensusRequestsType);
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Serialization.Rlp/ConsensusRequestDecoder.cs:45]: remove if no longer needed

File: src/Nethermind/Nethermind.Serialization.Rlp/ConsensusRequestDecoder.cs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                _depositDecoder.Encode(stream, (Deposit)item, rlpBehaviors);
+                break;
+        }
+    }
+
+    public Rlp Encode(ConsensusRequest item, RlpBehaviors rlpBehaviors = RlpBehaviors.None)
+    {
+        RlpStream rlpStream = new RlpStream(GetLength(item, rlpBehaviors));
+        // debug getLength
+        Console.WriteLine("GetLength: " + GetLength(item, rlpBehaviors));
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Serialization.Rlp/ConsensusRequestDecoder.cs:99]: Remove if no longer needed

File: src/Nethermind/Nethermind.Core.Test/Encoding/ConsensusRequestDecoderTests.cs [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+        codec.Encode(stream, withdrawalRequest);
+
+        Rlp.ValueDecoderContext decoderContext = new(stream.Data.AsSpan());
+        WithdrawalRequest? decoded = (WithdrawalRequest?)codec.Decode(ref decoderContext);
+
+        decoded.Should().BeEquivalentTo(withdrawalRequest);
+    }
+
+    [Test]
+    public void Should_encode_deposit_same_for_Rlp_Encode_and_DepositDecoder_Encode()
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Encoding/ConsensusRequestDecoderTests.cs:96]: we need tests for decoders many different types and for array of consensus requests

File: src/Nethermind/Nethermind.Serialization.Rlp/BlockDecoder.cs [hunk 15]
```diff
@@ -13,6 +14,8 @@ public class BlockDecoder : IRlpValueDecoder<Block>, IRlpStreamDecoder<Block>
         private readonly HeaderDecoder _headerDecoder = new();
         private readonly TxDecoder _txDecoder = new();
         private readonly WithdrawalDecoder _withdrawalDecoder = new();
+        private readonly DepositDecoder _depositDecoder = new();
+        private readonly WithdrawalRequestDecoder _WithdrawalRequestDecoder = new();
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Serialization.Rlp/BlockDecoder.cs:18]: Cosmetic: don't use capital letters for private vars

File: src/Nethermind/Nethermind.Core.Test/Encoding/BlockDecoderTests.cs [hunk 16]
```diff
@@ -86,6 +87,17 @@ public BlockDecoderTests()
...
+            Build.A.Block.WithNumber(1)
+                .WithBaseFeePerGas(1)
+                .WithTransactions(transactions)
+                .WithUncles(uncles)
+                .WithWithdrawals(8)
+                .WithBlobGasUsed(ulong.MaxValue)
+                .WithExcessBlobGas(ulong.MaxValue)
+                .WithMixHash(Keccak.EmptyTreeHash)
+                // an empty requests array
+                .WithConsensusRequests(0)
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Encoding/BlockDecoderTests.cs:100]: more RLP tests needed here: @rjnrohit

File: src/Nethermind/Nethermind.Core.Test/Encoding/DepositDecoderTests.cs [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+using Nethermind.Evm.Tracing.GethStyle.Custom.JavaScript;
+using Nethermind.Serialization.Rlp;
+using NUnit.Framework;
+
+namespace Nethermind.Core.Test.Encoding;
+
+public class DepositDecoderTests
+{
+    [Test]
+    public void Should_decode()
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Core.Test/Encoding/DepositDecoderTests.cs:17]: Is this class still needed @rjnrohit ? Seems to be duplicated with ConsensusRequestsDecoder?

File: src/Nethermind/Nethermind.State/Proofs/RequestsTrie.cs [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+        ArgumentNullException.ThrowIfNull(requests);
+    }
+
+    protected override void Initialize(ConsensusRequest[] requests)
+    {
+        var key = 0;
+
+        foreach (ConsensusRequest exit in requests)
+        {
+            Set(Rlp.Encode(key++).Bytes, _codec.Encode(exit).Bytes);
```
Comment (rjnrohit) [inline src/Nethermind/Nethermind.State/Proofs/RequestsTrie.cs:30]: Shouldn't it be `Set(Rlp.Encode(key++).Bytes, _codec.Encode(exit, RlpBehaviors.SkipTypedWrapping).Bytes);` ? @MarekM25

Comment (MarekM25) [reply src/Nethermind/Nethermind.State/Proofs/RequestsTrie.cs:30]: yes, I think so

File: src/Nethermind/Nethermind.Core.Test/Encoding/WithdrwalRequestDecoderTests.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+using Nethermind.Core.Test.Builders;
+using Nethermind.Serialization.Rlp;
+using NUnit.Framework;
+
+namespace Nethermind.Core.Test.Encoding;
+
+[TestFixture]
+public class WithdrawalRequestsDecoderTests
+{
+    private static WithdrawalRequestDecoder _decoder = new();
```
[the remark this answers is not in the document]
Comment (rjnrohit) [reply src/Nethermind/Nethermind.Core.Test/Encoding/WithdrwalRequestDecoderTests.cs:14]: Not needed

File: src/Nethermind/Nethermind.Serialization.Rlp/ConsensusRequestDecoder.cs [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+            ConsensusRequestsType.Deposit => _depositDecoder.GetContentLength((Deposit)item, rlpBehaviors),
+            _ => throw new RlpException($"Unsupported consensus request type {item.Type}")
+        };
+        return length;
+    }
+
+    public int GetLength(ConsensusRequest item, RlpBehaviors rlpBehaviors)
+    {
+        int length = item.Type switch
+        {
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Serialization.Rlp/ConsensusRequestDecoder.cs:27]: in GetLength we could call GetContentLength of ConsensusRequest to avoid extra switch here (check TxDecoder

File: src/Nethermind/Nethermind.Consensus.Clique/CliqueBlockProducer.cs [hunk 21]
```diff
@@ -445,7 +446,8 @@ bool IBlockProducer.IsProducingBlocks(ulong? maxProducingInterval)
             header,
             selectedTxs,
             Array.Empty<BlockHeader>(),
-            spec.WithdrawalsEnabled ? Enumerable.Empty<Withdrawal>() : null
+            spec.WithdrawalsEnabled ? Enumerable.Empty<Withdrawal>() : null,
+            spec.DepositsEnabled || spec.IsEip7002Enabled ? Enumerable.Empty<ConsensusRequest>() : null
```
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Consensus.Clique/CliqueBlockProducer.cs:450]: why ``spec.DepositsEnabled || spec.IsEip7002Enabled`` wouldn't make sense if it was ``spec.ConsensuRequestsEnabled`` and in IReleaseSpec we have the property there?

File: src/Nethermind/Nethermind.Core.Test/Builders/BlockBuilder.cs [hunk 22]
```diff
@@ -214,6 +215,13 @@ public BlockBuilder WithWithdrawalsRoot(Hash256? withdrawalsRoot)
             return this;
         }
 
+        public BlockBuilder WithDepositRoot(Hash256? depositsRoot)
```
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Core.Test/Builders/BlockBuilder.cs:218]: ig this one is obsolete no, right?

File: src/Nethermind/Nethermind.Core.Test/Builders/BlockHeaderBuilder.cs [hunk 23]
```diff
@@ -195,4 +195,10 @@ public BlockHeaderBuilder WithParentBeaconBlockRoot(Hash256? parentBeaconBlockRo
         TestObjectInternal.ParentBeaconBlockRoot = parentBeaconBlockRoot;
         return this;
     }
+
```
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Core.Test/Builders/BlockHeaderBuilder.cs:198]: this one too

File: src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs [hunk 24]
```diff
@@ -284,7 +299,46 @@ private bool ValidateWithdrawals(Block block, IReleaseSpec spec, out string? err
...
+        if ((spec.DepositsEnabled || spec.IsEip7002Enabled) && block.Requests is null)
+        {
+            error = BlockErrorMessages.MissingRequests;
+
+            if (_logger.IsWarn) _logger.Warn(error);
+
+            return false;
+        }
+
+        if (!spec.DepositsEnabled && !spec.IsEip7002Enabled && block.Requests is not null)
```
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs:316]:
```suggestion
        if (!spec.DepositsEnabled && !spec.WithdrawalRequestsEnabled && block.Requests is not null)
```

File: src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs [hunk 25]
```diff
@@ -284,7 +299,46 @@ private bool ValidateWithdrawals(Block block, IReleaseSpec spec, out string? err
         return true;
     }
 
-    private bool ValidateTransactions(Block block, IReleaseSpec spec, out string? errorMessage)
+    public bool ValidateRequests(Block block, out string? error) =>
+        ValidateRequests(block, _specProvider.GetSpec(block.Header), out error);
+
+    private bool ValidateRequests(Block block, IReleaseSpec spec, out string? error)
+    {
+        if ((spec.DepositsEnabled || spec.IsEip7002Enabled) && block.Requests is null)
```
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Consensus/Validators/BlockValidator.cs:307]:
```suggestion
        if ((spec.DepositsEnabled || spec.WithdrawalRequestsEnabled) && block.Requests is null)
```

File: src/Nethermind/Nethermind.Consensus/Requests/ConsensusRequestsProcessor.cs [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+    private readonly IDepositsProcessor _depositsProcessor;
+
+    public ConsensusRequestsProcessor()
+    {
+        _withdrawalRequestsProcessor = new WithdrawalRequestsProcessor();
+        _depositsProcessor = new DepositsProcessor();
+    }
+    public void ProcessRequests(IReleaseSpec spec, IWorldState state, Block block, TxReceipt[] receipts)
+    {
+        if (spec.IsEip6110Enabled == false && spec.IsEip7002Enabled == false)
```
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Consensus/Requests/ConsensusRequestsProcessor.cs:26]:
```suggestion
        if (spec.DepositsEnabled == false && spec.WithdrawalRequestsEnabled == false)
```

File: src/Nethermind/Nethermind.Consensus.Clique/CliqueBlockProducer.cs (hunk 21 above)
Comment (Demuirgos) [inline src/Nethermind/Nethermind.Consensus.Clique/CliqueBlockProducer.cs:450]:
```suggestion
            spec.DepositsEnabled || spec.WithdrawalRequestsEnabled ? Enumerable.Empty<ConsensusRequest>() : null
```

Comment (MarekM25) [reply src/Nethermind/Nethermind.Consensus.Clique/CliqueBlockProducer.cs]: yeah, wanted to do it, thx. Changed

**Review by LukaszRozmej — COMMENTED** [review]
Does ConsensusRequest need to be serialized to RLP (they are not in BlockBodies)?
Do they need to be transmitted through network? That would probably need an RLP change and new eth/XX version?
Do they need to be sent through RPC?

Basicly are they transient - just to reply once per NewPayload, or do we need to store them somehow?

**Comment by MarekM25** [comment]
> Does ConsensusRequest need to be serialized to RLP (they are not in BlockBodies)? Do they need to be transmitted through network? That would probably need an RLP change and new eth/XX version? Do they need to be sent through RPC?
> 
> Basicly are they transient - just to reply once per NewPayload, or do we need to store them somehow?

They are in block body, we haven't added them to BlockBodyMessageSerializer. @rjnrohit is working on it. They are not transient. They are similar to withdrawals. We need to store them. About RPC - I think they are part of block so it should be in RPC as well. I think we don't need a new version of eth/XX. They need to be transmitted through network

File: src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+    public byte[]? PubKeyField { get; set; }
+
+    [JsonIgnore]
+    public byte[]? WithdrawalCredentialsField { get; protected set; }
+
+    [JsonIgnore]
+    public byte[]? SignatureField { get; protected set; }
+
+    [JsonIgnore]
+    public ulong? IndexField { get; protected set; }
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs:36]: the classes Deposits and WithdrawalRequests are exposed in Engine API, that's why

File: src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecLoader.cs [hunk 28]
```diff
@@ -423,13 +435,25 @@ private static void LoadGenesis(ChainSpecJson chainSpecJson, ChainSpec chainSpec
             genesisHeader.ParentBeaconBlockRoot = Keccak.Zero;
         }
 
+        if (requestsEnabled)
+        {
+            genesisHeader.ReceiptsRoot = Keccak.EmptyTreeHash;
+        }
+
```
[the remark this answers is not in the document]
Comment (MarekM25) [reply src/Nethermind/Nethermind.Specs/ChainSpecStyle/ChainSpecLoader.cs:442]: receipts root is also initialized with EmptyTreeHash (a few lines above)

**Comment by rjnrohit** [comment]
> > Does ConsensusRequest need to be serialized to RLP (they are not in BlockBodies)? Do they need to be transmitted through network? That would probably need an RLP change and new eth/XX version? Do they need to be sent through RPC?
> > Basicly are they transient - just to reply once per NewPayload, or do we need to store them somehow?
> 
> They are in block body, we haven't added them to BlockBodyMessageSerializer. @rjnrohit is working on it. They are not transient. They are similar to withdrawals. We need to store them. About RPC - I think they are part of block so it should be in RPC as well. I think we don't need a new version of eth/XX. They need to be transmitted through network

Added!

File: src/Nethermind/Nethermind.Core/ConsensusRequests/WithdrawalRequest.cs [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public ulong Amount
+    {
+        get { return AmountField; }
+        set { AmountField = value; }
+    }
+    public override string ToString() => ToString(string.Empty);
+
+    public string ToString(string indentation) => new StringBuilder($"{indentation}{nameof(WithdrawalRequest)} {{")
```
[the remark this answers is not in the document]
Comment (rjnrohit) [reply src/Nethermind/Nethermind.Core/ConsensusRequests/WithdrawalRequest.cs:37]: done!

File: src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs (hunk 27 above)
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Core/ConsensusRequests/ConsensusRequest.cs:36]: Can we do it as a proper JSON class hierarchy? With `ConsensusRequestsType` being a discriminating field?

File: src/Nethermind/Nethermind.Core/Extensions/UInt64Extensions.cs [hunk 30]
```diff
@@ -0,0 +1,30 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+
+namespace Nethermind.Core.Extensions;
+
+public static class UInt64Extensions
+{
+    public static ulong ToULongFromBigEndianByteArrayWithoutLeadingZeros(this byte[]? bytes) =>
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Extensions/UInt64Extensions.cs:10]: if we go with Abi.UInt64 this won't be needed.
